main.cpp

```diff
@@ -0,0 +1,197 @@
+/** 
+ * @file pluviómetro.cpp
+ * @brief Proyecto pluviómetro
+ * 
+ * Este archivo contiene la implementación del proyecto pluviómetro 2 utilizando MBED.
+ */
+
+#include "mbed.h"
+#include "arm_book_lib.h"
+#include "weather_station.h"
+
+// Constantes de configuración
+#define TIME_INI  1593561600  ///< 1 de julio de 2020, 00:00:00
+#define BAUD_RATE 9600  ///< Velocidad de comunicación serial
+#define DELAY_BETWEEN_TICK 500  ///< 500 ms
+#define SWITCH_TICK_RAIN BUTTON1  ///< Botón para detectar lluvia
+#define RAINFALL_CHECK_INTERVAL 1  ///< Intervalo de verificación de lluvia en minutos
+#define MM_PER_TICK 0.2f  ///< 0.2 mm de agua por tick
+#define RAINFALL_COUNT_INI 0  ///< Contador de lluvia inicial
+#define LAST_MINUTE_INI -1  ///< Último minuto inicial
+
+// Mensajes y formatos
+#define MSG_RAIN_DETECTED " - Rain detected\r\n"  ///< Mensaje de lluvia detectada
+#define MSG_ACCUMULATED_RAINFALL " - Accumulated rainfall: "  ///< Mensaje de lluvia acumulada
+#define TIME_FORMAT "%Y-%m-%d %H:%M:%S"  ///< Formato de tiempo completo
+#define DATE_FORMAT "%Y-%m-%d %H:%M"  ///< Formato de fecha y hora
+
+// Sensores
+void initializeSensors();
+bool isRaining();
+
+// Análisis de Datos
+void analyzeRainfall();
+void accumulateRainfall();
+bool hasTimePassedMinutesRTC(int minutes);
+
+// Actuación
+void actOnRainfall();
+void reportRainfall();
+void printRain(const char* buffer);
+const char* DateTimeNow(void);
+void printAccumulatedRainfall();
+
+// Variables globales
+BufferedSerial pc(USBTX, USBRX, BAUD_RATE);  ///< Comunicación serial
+
+DigitalOut alarmLed(LED1);  ///< LED de alarma
+DigitalIn tickRain(SWITCH_TICK_RAIN);  ///< Botón de detección de lluvia
```

Comment:
No me queda claro el funcionamiento de este botón
Entiendo que este botón simula el funcionamiento del balancín? o estoy equivocado?
Si es así está mal usado. Debería detectar el paso del balancín de alguna manera

Reply:
si el botón simula el reed switch que esta en el balancín

Reply:
Entonces, según alcanzo a entender, cada vez que el balancín pasa de lado a lado, genera un pulso (00000110000) que es lo que hay que detectar:
https://youtu.be/sY0zeF0pZJ0
Esto significa un cambio significativo en el código

Reply:
Puede ser que no queda tan claro porque trabaje con la lógica propuesta de detectar -> procesar -> actuar. En este caso tickRain se ocupa para detectar la presencia de lluvia como se muestra a continuación:

/**
 * @brief Verifica si está lloviendo
 * 
 * @return true si el botón de detección de lluvia está activado, false en caso contrario
 */
bool isRaining() {
    return (tickRain == ON);
}

Luego desencadena una serie de eventos con la función actOnRainfall();
que es la encargada de ejecutar  la función encargada de "analizar" el evento de precipitación  analyzeRainfall();
Esto contempla disparar un evento de detección de lluvia e incrementar un contador de precipitación que en este caso es de 1 minuto (generalmente esta seteado en 1 hora) . 
Creo que haré un esquema para que quede mas claro todo

Reply:
https://github.com/cpatagon/TP1_pluviometro/blob/08465128de7a988033c8095deff73ed6353271af/README.md?plain=1#L29C1-L36C59

Reply:
El pluviómetro del diagrama: doc/fig/pluviometro.png funciona de la siguiente manera:
_El agua de lluvia es conducida desde un colector balancín con dos cangilones calibrados. La lluvia cae en un cangilón que se va llenando y una vez alcanzado un volumen deseado, el peso del agua hace que bascule y enfrente el siguiente cangilón al orificio de salida de agua. La lluvia caída estará relacionada con el número de oscilaciones del balancín y la relación dependerá de la capacidad del cangilón. Estas oscilaciones son detectadas por un sensor y enviadas a un microprocesador._
No encuentro relación alguna entre este diagrama, el README y el esquema, con el código.
Deberías detectar el pasaje del cangilón (con el reedswitch) y sumar la capacidad del cangilón al acumulado de lluvia

Reply:
Hola @EFilomena,

Te recomiendo mirar la figura 3 en este enlace: [Diagrama del pluviómetro](https://github.com/cpatagon/TP1_pluviometro/blob/revision_TP1/doc/fig/Diagrama_pluviometro.png).

El botón denominado 
![Uploading Diagrama_pluviometro.png…]()
tickRain es el componente equivalente al reedswitch que detecta la lluvia (en estado ON). Este interruptor se encuentra dentro de la función isRaining(), la cual está dentro del bucle infinito, en el if. Cuando se detecta un cambio en la posición del balancín, se ejecuta la función actOnRainfall(), que incrementa el contador en accumulateRainFall(). Los datos se acumulan durante un minuto y luego se imprime el número de ticks, expresado en milímetros de agua. Cada tick equivale a 0.02 mm. Después, el contador de lluvia se reinicia a 0 y se espera un nuevo tick.

Concuerdo en que al diagrama le falta el reinicio del contador. También estoy de acuerdo en que el tick debería ser manejado mediante una interrupción, pero creo que esto se verá más adelante.

Reply:
ya se agrego en la figura 3, del diagrama, el reinicio del contador -> rainfallCount = RAINFALL_COUNT_INI; tambien esta el sumador de lluvia rainfallCount ++ que se reinicia cada 1 minuto (únicamente durante las pruebas) ya que lo normal son 1 hora o un día

Reply:
El problema es que el pluviómetro propuesto en la figura (doc/fig/pluviometro.png) no funciona como vos decís. No devuelve un "ON" cuando está lloviendo y un "OFF" cuando no está lloviendo. No funciona así.

Reply:
@cpatagon Lo que Eduardo te dice es que estás presentando un trabajo en donde decís que hay un balancín y no estás mirando el balancín.
Lo que yo entiendo que vos decís es que la simple presión del botón indica la presencia de lluvia.

Si estoy en lo cierto, toda la referencia y explicación sobre el postprocesamiento no aportan nada a la discusión, y tal vez con un simple "tenés razón, pero ahora es el TP1 y quiero hacerlo simple", todos podemos emplear mejor nuestro tiempo.
Evidentemente estás observando que un pin se ponga en 1, sin antirebote, y sin observar un segundo pin; lo cual refuerza mi presunción de certeza.

Cada vez que el botón rebota caen tantos pulsitos de lluvia como rebotes.

Cada vez que alguien enciende un tubo fluorescente de los viejos, contás lluvia. SI hay tormenta eléctrica, contás lluvia.

Reply:
encontré esta imagen que puede ayudar  
![imagen](https://github.com/user-attachments/assets/cdd79982-f6b4-40d1-b4b4-aad73d23825d)